raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,298 @@
+import "./TokenNetwork.sol";
+import "./RaidenServiceBundle.sol";
+
+contract MonitoringService is Utils {
+
+    // Token to be used for paying the rewards
+    Token public token;
+
+    // Raiden Service Bundle contract to use for checking if MS has deposits
+    RaidenServiceBundle public rsb;
+
+    // keccak256(channel_identifier, token_network_address) => Struct
+    // Keep track of the rewards per channel
+    mapping(bytes32 => Reward) rewards;
+
+    // Keep track of balances
+    mapping(address => uint) public balances;
+
+    /*
+     *  Structs
+     */
+    struct Reward{
+        // The amount of tokens to be rewarded
+        uint256 reward_amount;
+
+        // Nonce of the most recently provided BP
+        uint256 nonce;
+
+        // Address of the Raiden Node that was monitored
+        // This is also the address that has the reward deducted from its deposit
+        address reward_sender_address;
+
+        // Address of the Monitoring Service who is currenly eligible to claim the reward
+        address monitoring_service_address;
+    }
+
+    /*
+     *  Events
+     */
+
+    event NewDeposit(address receiver, uint amount);
+    event NewBalanceProofReceived(
+        uint256 reward_amount,
+        uint256 nonce,
+        address ms_address,
+        address raiden_node_address
+    );
+    event RewardClaimed(address ms_address, uint amount, bytes32 reward_identifier);
+    event Withdrawn(address account, uint amount);
+
+    /*
+     *  Modifiers 
+     */
+
+    modifier canMonitor(address _ms_address) {
+        require(rsb.deposits(_ms_address) > 0);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    /// @param _rsb_address The address of the RaidenServiceBundle contract
+    constructor(
+        address _token_address,
+        address _rsb_address
+    )
+        public
+    {
+        require(_token_address != 0x0);
+        require(_rsb_address != 0x0);
+        require(contractExists(_token_address));
+        require(contractExists(_rsb_address));
+
+        token = Token(_token_address);
+        rsb = RaidenServiceBundle(_rsb_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+        // Check if the contract is indeed an rsb contract
+        // TODO: Check that some function exists in the contract
+    }
+
+    /// @notice Deposit tokens used to either pay MSs or to regsiter as MS
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefitting from the deposit
+    /// @param amount The amount of tokens to be depositted
+    function deposit(address beneficiary, uint amount) public {
+        require(amount > 0);
+
+        // This also allows for MSs to deposit and use other MSs
+        balances[beneficiary] += amount;
```

Comment:
In the TokenNetwork contract we have decided to use an idempotent `setDeposit` function, which receives the total amount of tokens that should end up as deposited in the contract, instead of the amount that is added when calling the function. The main reason for this was to ensure that you don't `deposit` tokens by mistake, due to an additional API call.
I think we should be consistent and also use this approach here.